src/execution_plan/ops/op_procedure_call.c

```diff
@@ -87,47 +98,63 @@ OpBase *NewProcCallOp(const ExecutionPlan *plan, const char *procedure, const ch
 	return (OpBase *)op;
 }
 
-static OpResult ProcCallInit(OpBase *opBase) {
-	OpProcCall *op = (OpProcCall *)opBase;
-	ProcedureResult res = Proc_Invoke(op->procedure, op->args);
-	return (res == PROCEDURE_OK) ? OP_OK : OP_ERR;
-}
-
 static Record ProcCallConsume(OpBase *opBase) {
 	OpProcCall *op = (OpProcCall *)opBase;
-	Record r = NULL;
-
-	if(op->op.childCount == 0) {
-		/* Make record large enough to accommodate all alias entities. */
-		r = OpBase_CreateRecord((OpBase *)op);
-	} else {
-		OpBase *child = op->op.children[0];
-		r = OpBase_Consume(child);
-		if(!r) return NULL;
-	}
 
-	SIValue *outputs = Proc_Step(op->procedure);
-	if(outputs == NULL) {
-		Record_Free(r);
-		return NULL;
+	Record yield_record = NULL;
+	while(!(yield_record = _yield(op))) {
+		// Free old record.
+		if(op->r) {
+			Record_Free(op->r);
+			op->r = NULL;
+		}
+
+		if(op->op.childCount == 0) {
+			// "Static evaluation", return data for first call only!
+			if(!op->first_call) return NULL;
+			op->first_call = false;
+			op->r = OpBase_CreateRecord((OpBase *)op);
+		} else {
+			OpBase *child = op->op.children[0];
+			op->r = OpBase_Consume(child);
+			if(!op->r) return NULL;
+		}
+
+		// Evaluate arguments, free args from previous call.
+		for(uint i = 0; i < op->arg_count; i++) SIValue_Free(op->args + i);
```

Comment:
On the first call this free is unsafe, as this array is instantiated but there are no SIValues in it.
Postpone toggling `op->first_call` or free based on `array_len` rather than `op->argc`?

Reply:
👍